hal/src/main/native/athena/HAL.cpp

```diff
@@ -224,14 +224,23 @@ uint64_t HAL_GetFPGATime(int32_t* status) {
     *status = NiFpga_Status_ResourceNotInitialized;
     return 0;
   }
-  std::lock_guard<std::mutex> lock(timeMutex);
-  uint32_t fpgaTime = global->readLocalTime(status);
+  uint64_t upper1 = global->readLocalTimeUpper(status);
+  uint32_t lower = global->readLocalTime(status);
+  uint64_t upper2 = global->readLocalTimeUpper(status);
   if (*status != 0) return 0;
-  // check for rollover
-  if (fpgaTime < prevFPGATime) ++timeEpoch;
-  prevFPGATime = fpgaTime;
-  return static_cast<uint64_t>(timeEpoch) << 32 |
-         static_cast<uint64_t>(fpgaTime);
+  constexpr uint32_t rolloverCheckValue = 0xFFFFFF00;
+  if (lower < rolloverCheckValue)
```

Comment:
Since you're always reading upper twice, you can just check to see if it changed (e.g. upper1 != upper2) instead of comparing the rollover value.

Reply:
Well if that was the case, you could always just use upper2. That would give different behavior then the comparison. After some more thought, this would actually require a 4th read, but we could conditionally do that, since it would only happen on the single rollover case, so the jitter would be nothing.